### PR TITLE
Add Powered rails

### PR DESCRIPTION
**Pull request by Hamms**
This PR does a bit more work than it needs to do, but it's laying a fair amount of groundwork for the redstone refactor I'm working on.

Notably, it [adds the concept of Adjacency Sets](https://github.com/code-dot-org/craft/commit/2302148aff5b09b740bdbd0dd51e83be34bb6167) to the Position Helper. This specific implementation of adjacency sets should be considered at least partially a proof of concept; it's significantly more computationally complex than it should be, and I intend to improve that.

With that done, the [addition of a `powerRails` step](https://github.com/code-dot-org/craft/commit/71797c056ab3ebe8cb29207b735504b650459023) simply uses a custom definition of "adjacency" to gather connectivity sets of rails.

### CODE REVIEW

File: src/js/game/LevelMVC/LevelModel.js [hunk 1]
```diff
@@ -357,19 +357,6 @@ module.exports = class LevelModel {
...
-      for (var y = 0; y < this.planeHeight; ++y) {
-        var block = this.actionPlane.getBlockAt([x, y]);
-        if (block.blockType.substring(0, 7) === "railsUn") {
-          unpoweredRails.push([x, y], "railsPowered" + block.blockType.substring(14));
-        }
-      }
-    }
-    return unpoweredRails;
-  }
-
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelModel.js]: 🎉

File: src/js/game/LevelMVC/LevelPlane.js [hunk 2]
```diff
@@ -253,6 +253,86 @@ module.exports = class LevelPlane {
...
   /**
+   * Determine whether or not the blocks at the given positions are powered
+   * rails that are connected to each other
+   *
+   * @param {Posititon} left
+   * @param {Posititon} right
+   * @return {boolean}
+   */
+  getPoweredRailsConnected(left, right) {
+    // return early if they're not even adjacent
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelPlane.js]: Nit: `// Return early if the positions are not event adjacent.`

File: src/js/game/LevelMVC/LevelPlane.js [hunk 3]
```diff
@@ -253,6 +253,86 @@ module.exports = class LevelPlane {
   }
 
   /**
+   * Determine whether or not the blocks at the given positions are powered
+   * rails that are connected to each other
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelPlane.js]: Tiny nit: missing punctuation.

**Comment by joshlory** [comment]
I think this change may regress https://github.com/code-dot-org/craft/issues/105?  Specifically, this (skipped) test passes on master but fails on this branch: https://github.com/code-dot-org/craft/blob/807a39627c36f384c1b33edd0441d57af6618b67/test/unit/LevelPlaneTest.js#L245

File: src/js/game/LevelMVC/LevelPlane.js [hunk 4]
```diff
@@ -253,6 +253,86 @@ module.exports = class LevelPlane {
...
+    // to be connected, both blocks must be powerable rails
+    if (!(leftBlock.getIsPowerableRail() && rightBlock.getIsPowerableRail())) {
+      return false;
+    }
+
+    // to be connected, both blocks must be oriented either North/South or
+    // East/West
+    const bothEastWest =
+      leftBlock.blockType.match('East|West') &&
+      rightBlock.blockType.match('East|West');
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelPlane.js]: Worth adding a `rightBlock.isHorizontal()` helper method?  A bit gross now when all block types live in the same class, but might make sense in the future when the different block types are separate classes.

**Comment by Hamms** [comment]
Huh; why is that test getting skipped?

**Comment by joshlory** [comment]
Never mind!  I had pulled a local `powered-rails` branch that happened to be named the same as your PR branch 😄.

**Comment by Hamms** [comment]
That test is passing on this branch for me ... can you confirm your repro?

File: src/js/game/LevelMVC/LevelPlane.js [hunk 5]
```diff
@@ -253,6 +253,86 @@ module.exports = class LevelPlane {
...
+      return false;
+    }
+  }
+
+  /**
+   * Propagate power to (and orient) all powerable rails in the level.
+   */
+  powerRails() {
+    // find all rails that can be powered
+    const powerableRails = this.getAllPositions().filter((position) => (
```
Comment (joshlory) [inline src/js/game/LevelMVC/LevelPlane.js]: Tiny nit: extra parens around `position`.

**Pull request merged**